scripts/dts/extract_dts_includes.py

```diff
@@ -18,6 +19,8 @@
 aliases = {}
 chosen = {}
 reduced = {}
+defs = {}
```

Comment:
@mbolivar: I can only ack your observations.
Agreed that current version the script is not a state of the art piece of python software but rather a simple script doing job expected from it. Aim of this series was primarily focused on adding functionality to it.
But it would indeed make lot of sense to increase a bit the standards on this script, as I'm sure it could simplify some processing, make it easier to use and more powerful. To do this though, I humbly admit I could use some help, as I'm still a python apprentice. Well, for now, I can at least provide more  comments on the code.

Before doing this, maybe one point would be to know where we're going and what functionality is expected from the script. If we agree that long term goal is to generate code that could be used for driver initialisation (driver instances and their configuration data), there are several ways to do that.
I see two trends currently:
 - Initial and current 'defs' trend, generating pure #defines with machine computable names (which is mostly available today). Then post process these definitions so driver can get rid of Kconfig stuff related to instances/config/... This part is currently an unknown (to me at least)
 - New 'struct' trend introduced here, that allows to store information from dts/yaml in a struct dict holding parsed information and hence directly ready to generate whatever code is needed (an example of use of such struct dict is here: https://github.com/erwango/zephyr/commits/dts_struct_cleanup generating directly uart driver init code per instance, skipping the #define generation part )

If a rework is required, this should be decided first.
@galak, @nashif, @agross-linaro : maybe you have something in mind already?

Reply:
> Before doing this, maybe one point would be to know where we're going and what functionality is expected from the script.

@erwango I agree this is the main point.

My concern is Zephyr relies on DT more and more (e.g. increasingly complex code generation in this series), but has an important DT script that is hard to understand and modify.

If this is the "last" feature for this script, maybe OK, but if we're going farther from here, this seems worrisome.

> But it would indeed make lot of sense to increase a bit the standards on this script, as I'm sure it could simplify some processing, make it easier to use and more powerful. To do this though, I humbly admit I could use some help, as I'm still a python apprentice. Well, for now, I can at least provide more comments on the code.

Totally fair, and I want to repeat I'm not trying to block this series with my comments.

Reply:
> Totally fair, and I want to repeat I'm not trying to block this series with my comments.

No problem, this RFC is meant to be discussed. DTS based code generation will have some impacts throughout zephyr code, it deserves some discussion beforehand.

Reply:
> No problem, this RFC is meant to be discussed. DTS based code generation will have some impacts throughout zephyr code, it deserves some discussion beforehand.

On that note: looking around for Python representations of device trees, I was surprised and pleased to see that Zephyr developer @superna9999 already has a library for managing flattened device trees in Python!

https://github.com/superna9999/pyfdt

It has a great looking set of classes for representing nodes and properties which has thorough documentation:

https://github.com/superna9999/pyfdt/blob/master/pyfdt/pyfdt.py

@superna9999, do you think this is something Zephyr could easily use instead of its current scripts for managing DTS files? I know FDT is of course a different format, but perhaps it's not so hard to either have Zephyr generate a DTB, or perhaps have the library parse DTS?

Reply:
@mbolivar Hi, thanks for pointing it !

Actually it was already evaluated in the early days on DT in zephyr, but since pyfdt does not have a dts compiler, it needed a painful 'dtc' compilation to dtb then an import into pyfdt.

But then a simple dts parser was written and pyfdt was ditched, but then I always had in mind to use this parser as a pyfdt frontend to compile dts files... even the licence matches !

It could be a good base to import in zephyr code base as a foundation to a brand new code generator from DT.

Reply:
@mbolivar , @superna9999, I had a quick look to pyfdt.

While I still agree on the observation that current script should be deeply reworked or replaced for an easier maintenance and future evolutions, I'm not sure pyfdt is the good match, from a functional point of view, IMVHO.
It is definitely a good example, and I don't have better alternative to propose, but I think pyfdt provides a lot of functionalities but only a little that would be useful for code generation once dts is already available.
I'm not convinced we need to compile dts file to be able to parse it. devicetree.py converts .dts into a python dict object that is pretty sufficient for further parsing. Code generation part (what would need to be enhanced, ie 90% of the script) happens after this.

Reply:
> It is definitely a good example, and I don't have better alternative to propose, but I think pyfdt provides a lot of functionalities but only a little that would be useful for code generation once dts is already available.
> I'm not convinced we need to compile dts file to be able to parse it. devicetree.py converts .dts into a python dict object that is pretty sufficient for further parsing. Code generation part (what would need to be enhanced, ie 90% of the script) happens after this.

I don't have plans to contribute heavily to this effort so I will happily defer to your opinion.

I would only say that I think the pyfdt library seems to provide good class definitions with good documentation and decent duck-typing for some of the key properties DT seems to rely on. This is a very beginner opinion, and so it's likely I've missed many important details.

I think the nested dicts that Zephyr is using now are very hard to understand, and very hard to typecheck, even at runtime. But this is not necessarily a problem if only a few people who all know what they are doing are interacting with the datatypes.

Reply:
IMHO converting to dict will be limiting at some point, having a proper object model will get the code maintainable and understandable against the device tree standards.
Copying pyfdt may be a bad idea, but implementing a similar model maybe smart towards to goals we want to achieve with device tree support.
Basically, we are trying to do what is actually done in the linux kernel, in python, at compile time with a limited object representation.
At some point, we should restructure this script into submodules mapping each bindings, i.e. pinctrl, interrupt parsing, i2c bus, spi bus, flash partitions... from a correctly parsed device tree.

Reply:
@superna9999 , @mbolivar : Sorry, I had missed your comments

> IMHO converting to dict will be limiting at some point, having a proper object model will get the code maintainable and understandable against the device tree standards.

Ok, I will trust your experience on this

> implementing a similar model maybe smart towards to goals we want to achieve with device tree support.

For sure

> At some point, we should restructure this script into submodules mapping each bindings, i.e. pinctrl, interrupt parsing, i2c bus, spi bus, flash partitions... from a correctly parsed device tree.

On this point I'm not so sure. I think (hope) yaml files should provide enough information to have a nearly generic parsing so it's not needed to add a new submodule for each new binding. At least most people should be able to get things working by only adding a yaml file matching their binding. I know this is not 100% feasible, but I think this should be a target.